compiler/rustc_builtin_macros/src/format.rs

```diff
@@ -347,21 +117,19 @@ fn parse_args<'a>(
                 p.bump();
                 p.expect(&token::Eq)?;
                 let e = p.parse_expr()?;
-                if let Some(&prev) = names.get(&ident.name) {
-                    ecx.struct_span_err(e.span, &format!("duplicate argument named `{}`", ident))
-                        .span_label(args[prev].expr.span, "previously here")
-                        .span_label(e.span, "duplicate argument")
-                        .emit();
+                if let Some(prev) =
+                    args.iter().rev().map_while(|a| a.1.ident()).find(|n| n.name == ident.name)
```

Comment:
Shouldn't we instead "waste" some memory keeping a `HashSet<Symbol>` or `HashSet<Ident>` and checking that here? This is O(n^2) *in the happy path* for a case that I'm pretty sure isn't tracked in the perf suite.

Reply:
Maybe. A linear search through a short Vec is extremely fast though. I'm not sure how often we'd have a `format_args!()` so big that the `O(..)` behavior becomes relevant.

Reply:
It's not so much about how bad it is in the general case, it is about resiliency in the extremes. I could see for example a templating engine ending up being a big `format` call with a ton of arguments.

Reply:
Ah right, that makes sense.

I'll change the `Vec<(P<Expr>, FormatArgKind)>` to a new struct that represents the arguments in a way that allows quick lookup through a hashmap. (That also resolves your comment about `start_of_named_args` below.)

(Not sure what to name that struct though. :thinking: `Arguments`? `FormatArguments`? `ArgumentsToFormatArgs`? `FormatArgsArguments`? `FormatArgsArgs`? Confusing.)

Reply:
Updated.